finality-aleph/src/data_io/data_store.rs

```diff
@@ -0,0 +1,636 @@
+        };
+        let new_status: ProposalStatus =
+            self.check_proposal_availability(proposal, Some(&old_status));
+        self.pending_proposals.get_mut(proposal).unwrap().status = new_status.clone();
+
+        use crate::data_io::proposal::ProposalStatus::*;
+        match new_status {
+            PendingTopBlock => {
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            TopBlockImportedButIncorrectBranch => {
+                // We do nothing, this is a result of malicious behaviour. This proposal can still get through,
+                // but we wait till maintenance.
+                false
+            }
+            TopBlockImportedButNotFinalizedAncestor => {
+                // This is possible in honest scenarios -- this node must be behind in finalization.
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            Available(_) => {
+                self.on_proposal_available(proposal);
+                true
+            }
+        }
+    }
+
+    // Outputs the current status of the proposal based on the `old_status` (for optimization).
+    fn check_proposal_availability(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: Option<&ProposalStatus>,
+    ) -> ProposalStatus {
+        self.proposal_status_provider
+            .lock()
+            .get_status(proposal, old_status)
+    }
+
+    // For a proposal that might be new or not, check if it is available. If it is a new proposal
+    // and it is not available:
+    // 1) create an entry in pending_proposals
+    // 2) link this proposal to the input message by modifying the new added entry in pending proposals
+    //    and the provided message_info.
+    // 3) register an appropriate event trigger (for block import and/or finality).
+    // If the proposal is available, message_info is not modified.
+    fn add_message_proposal_dependency(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        message_info: &mut PendingMessageInfo<B, Message>,
+        id: MessageId,
+    ) {
+        if !self.pending_proposals.contains_key(proposal) {
+            use crate::data_io::proposal::ProposalStatus::*;
+            let status = self.check_proposal_availability(proposal, None);
+            match status {
+                PendingTopBlock => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_block_import_trigger(proposal, &proposal.top_block());
+                    self.register_next_finality_trigger(proposal);
+                }
+                TopBlockImportedButIncorrectBranch => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_next_finality_trigger(proposal);
```

Comment:
The only way this might ever get through is as a hopeless fork. So the only event that might change the status of this proposal is a finalization event. Added a comment